src/policies.rs

```diff
@@ -375,6 +443,79 @@ impl AllocBlock {
             .correct_buffer()
     }
 
+    pub fn encrypt_buffer_ffi<'a>(w: WeakMut<'a, AllocBlock>) {
+        w.get_ref_mut()
+            .expect("encrypt_buffer_ffi")
+            .encrypt_buffer()
+    }
+
+    pub fn decrypt_buffer_ffi<'a>(w: WeakMut<'a, AllocBlock>) {
+        w.get_ref_mut()
+            .expect("decrypt_buffer_ffi")
+            .decrypt_buffer()
+    }
+
+    /// TODO: refactor encrypt and decrypt buffer
+    fn encrypt_buffer(&mut self) {
+        let mut buffer = self.buffer();
+
+        match self.policies.iter().position(|&pol| pol.is_red()) {
+            Some(idx) => {
+                buffer = self.policies[idx].get_data_mut(buffer);
+            }
+            None => (),
+        }
+
+        match self.policies.iter().position(|&pol| pol.is_rs()) {
+            Some(idx) => {
+                buffer = self.policies[idx].get_data_mut(buffer);
+            }
+            None => (),
+        }
+
+        match self.policies.iter().position(|&pol| pol.is_crypt()) {
+            Some(idx) => {
+                let key = GenericArray::from_slice(KEY);
+                let nonce = GenericArray::from_slice(NONCE);
+                let mut cipher = Aes128Ctr::new(&key, &nonce);
+                let (mut data, err) = self.policies[idx].split_buffer_mut(buffer); 
```

Comment:
Won't `split_buffer_mut` be wrong here if say the policy was read solomon? Because `get_data_mut`  already calls `split_buffer_mut` so I think you double split

Reply:
No, this is only for the case where the policy is in fact `Policy::Encrypted`. That's part of the lambda in the `match`.